### PR TITLE
core/state: implement fast storage deletion

### PR DESCRIPTION
**Pull request by rjl493456442**
```
Deletion of 328239 slots 22.47 MiB statesize in time 812.148619ms
800ms vs 3.6 s
```

### CODE REVIEW

**Comment by holiman** [comment]
Looking at the old code now.. Ok, so for `hashdb`, it first iterates the storage trie,  collects it into the `set` var: `aborted, slots, set, err := s.deleteStorage(addr, addrHash, prev.Root)`, 

The `set` are merged into `nodes`
```
		if err := nodes.Merge(set); err != nil {
			return nil, err
		}

```
Other changes are merged into the same `nodes`, and eventually, we update the triedb:

```
		if err := s.db.TrieDB().Update(root, origin, block, nodes, triestate.New(s.accountsOrigin, s.storagesOrigin, incomplete)); err != nil {
			return common.Hash{}, err
		}
```
And, eventually, it reaches `Update`, and does
```
	for _, owner := range order {
		subset := nodes.Sets[owner]
		subset.ForEachWithOrder(func(path string, n *trienode.Node) {
			if n.IsDeleted() {
				return // ignore deletion
			}

```
So it just ignores all deletions in the end. So all the work we did there was just discarded in the end? Or was there some period where it was used in a memory-capacity, and we just discard the actual disk-write? 

If indeed we just iterate/collect the storage nodes just to ignore them in hash-mode, it seems that we can optimize that particular pipeline earlier, and ignore the iteration too?

**Comment by holiman** [comment]
> I added this logic in this PR.

Right, gotcha! I'm a couple of steps behind here :)

**Comment by rjl493456442** [comment]
> It looks like the only reason for the first collection is to filter out the empty-element, but that can be easily done in the second loop. Am I missing something?

Yep, basically we need to flush storage trie(s) first, and then account trie. I think logically the refactor is OK, but i would prefer to not change it, because last time I changed the logic here and result in a big bug in this part :)

**Comment by rjl493456442** [comment]
Running a full sync on sepolia to ensure nothing is broken.

File: core/state/statedb.go [hunk 1]
```diff
@@ -1022,59 +1028,124 @@ func (s *StateDB) clearJournalAndRefund() {
...
+		size += common.StorageSize(common.HashLength + len(slot))
+		slots[iter.Hash()] = slot
+
+		if err := stack.Update(iter.Hash().Bytes(), slot); err != nil {
+			return false, 0, nil, nil, err
+		}
+	}
+	if stack.Hash() != root {
+		return false, 0, nil, nil, fmt.Errorf("snapshot is not matched, exp %x, got %x", root, stack.Hash())
+	}
```
Comment (holiman) [inline core/state/statedb.go:1065]:
```suggestion
	if it.Err() != nil {
		return false, 0, nil, nil, it.Err()
	}
	if stack.Hash() != root {
		return false, 0, nil, nil, fmt.Errorf("snapshot is not matched, exp %x, got %x", root, stack.Hash())
	}
```

Comment (holiman) [reply core/state/statedb.go:1065]: In theory, both `it.Next()` and `it.Hash()` can internally error, so we _should_ check `it.Err()` after each invocation. If we fail to do so, and ignore an error from `it.Hash()`, and subsequently call `it.Next()`, then we will be rewarded with `panic(fmt.Sprintf("called Next of failed iterator: %v", it.fail))`

But I guess it should be enough to have one check after the loop, like my comment, and another check  right after `slots[iter.Hash()] = slot` ?

Comment (rjl493456442) [reply core/state/statedb.go:1065]: Technically, we need to check the failure after each iteration, however, i guess it's unnecessary. The cached error won't be clean anyway.

And the reason I don't add the iterator error checking here is: if we encounter any failure in iterator, the stackTrie will produce a different root hash anyway.

But yeah, I will add one check right after the loop. It's cheap anyway and would be helpful to bubble up the "real issue".

**Review by holiman — APPROVED** [review]
I added a test aswell -- maybe it was superfluous, I couldn't really tell. Anyway, LGTM!

File: core/state/statedb.go (hunk 1 above)
Comment (holiman) [reply core/state/statedb.go:1068]: > And the reason I don't add the iterator error checking here is: if we encounter any failure in iterator, the stackTrie will produce a different root hash anyway.

Ah, no, as I pointed out, we won't get a different root, we will encounter a panic crash.

**Pull request merged**